src/main/scala/io/qbeast/spark/delta/writer/RollupWriteStrategy.scala

```diff
@@ -0,0 +1,167 @@
+      val row = InternalRow.fromSeq(extractors.map(_.apply(indexedRow))).copy()
+      val cubeIdBytes = indexedRow.getBinary(qbeastColumns.cubeColumnIndex)
+      val cubeId = tableChanges.updatedRevision.createCubeId(cubeIdBytes)
+      val weight = Weight(indexedRow.getInt(qbeastColumns.weightColumnIndex))
+      val rollupCubeIdBytes = indexedRow.getBinary(qbeastColumns.cubeToRollupColumnIndex)
+      val rollupCubeId = tableChanges.updatedRevision.createCubeId(rollupCubeIdBytes)
+      (row, cubeId, weight, rollupCubeId)
+    }
+  }
+
+  private def getRollupCubeIdUDF(rollupCubeIds: Map[CubeId, CubeId]): UserDefinedFunction =
+    udf((cubeIdBytes: Array[Byte]) => {
+      val cubeId = tableChanges.updatedRevision.createCubeId(cubeIdBytes)
+      var rollupCubeId = rollupCubeIds.get(cubeId)
+      var parentCubeId = cubeId.parent
+      while (rollupCubeId.isEmpty) {
+        parentCubeId match {
+          case Some(value) =>
+            rollupCubeId = rollupCubeIds.get(value)
+            parentCubeId = value.parent
+          case None => rollupCubeId = Some(cubeId)
+        }
+      }
+      rollupCubeId.get.bytes
+    })
+
+  private def computeRollupCubeIds: Map[CubeId, CubeId] = {
+    val minRowsPerFile = tableChanges.updatedRevision.desiredCubeSize.toDouble
+    val queue = new mutable.PriorityQueue()(CubeIdOrdering)
+    val rollups = mutable.Map.empty[CubeId, Rollup]
+    tableChanges.cubeDomains.foreach { case (cubeId, domain) =>
+      queue += cubeId
+      val minWeight = getMinWeight(cubeId).fraction
+      val maxWeight = getMaxWeight(cubeId).fraction
+      val size = (maxWeight - minWeight) * domain
+      rollups += cubeId -> Rollup(cubeId, size)
+    }
+
+    while (queue.nonEmpty) {
+      val cubeId = queue.dequeue()
+      val rollup = rollups(cubeId)
+      if (rollup.size < minRowsPerFile) {
+        cubeId.parent match {
+          case Some(parentCubeId) =>
+            val parentRollup = rollups.get(parentCubeId) match {
+              case Some(value) => value
+              case None =>
+                queue += parentCubeId
+                val value = Rollup(parentCubeId, 0)
+                rollups += parentCubeId -> value
+                value
+            }
+            parentRollup.append(rollup)
+            rollups -= cubeId
+          case None => ()
+        }
+      }
+    }
+
+    rollups.flatMap { case (cubeId, rollup) =>
+      rollup.cubeIds.map((_, cubeId))
+    }.toMap
+  }
+
+  private def getMinWeight(cubeId: CubeId): Weight = {
+    cubeId.parent match {
+      case Some(parentCubeId) => getMaxWeight(parentCubeId)
+      case None => Weight.MinValue
+    }
+  }
+
+  private def getMaxWeight(cubeId: CubeId): Weight = {
+    tableChanges.cubeWeight(cubeId).getOrElse(Weight.MaxValue)
+  }
+
+  private object CubeIdOrdering extends Ordering[CubeId] {
+    // Cube identifiers are compared by depth in reversed order.
+    override def compare(x: CubeId, y: CubeId): Int = y.depth - x.depth
```

Comment:
If the idea is to do rollup from the bottom up, is it possible that it should be `x.depth - y.depth`?